### PR TITLE
Implement db healing of broken database

### PR DESCRIPTION
**Pull request by thaarok**
Note: Carmen version in this PR needs to be replaced with the main branch when available.

Usage:
```
sonicd --config=config.toml --datadir=/tmp/operatest-break --fakenet=0/2 --port 5051 --http --http.api=eth,debug,net,admin,web3,personal,txpool,ftm,dag --http.corsdomain * --verbosity=3 --nat extip:127.0.0.1
WARN [08-01|21:06:20.869] Please add '--cache 11968' flag to allocate more cache for Opera. Total memory is 23937 MB. 
INFO [08-01|21:06:20.870] Maximum peer count                       total=50
failed to initialize the node: failed to make consensus engine: failed to open existing databases: dirty state: gossip: DE

sonictool --datadir=/tmp/operatest-break --cache 11968 heal
INFO [08-01|21:09:11.489] Maximum peer count                       total=50
INFO [08-01|21:09:11.489] Forcing unlock of directory              dir=/tmp/operatest-break/carmen/archive
INFO [08-01|21:09:11.489] Healing gossip db... 
INFO [08-01|21:09:11.495] State for the last closed epoch is not available epoch=20
INFO [08-01|21:09:11.495] Last closed epoch with available state found epoch=19
INFO [08-01|21:09:11.495] Reverting to epoch state                 epoch=19 block=968
INFO [08-01|21:09:11.496] Removing excessive events 
INFO [08-01|21:09:11.497] Removing epoch DBs - will be recreated on next start 
INFO [08-01|21:09:11.497] Removing epoch db                        name=gossip-20
INFO [08-01|21:09:11.502] Removing epoch db                        name=lachesis
INFO [08-01|21:09:11.508] Removing epoch db                        name=lachesis-20
INFO [08-01|21:09:11.513] Recreating consensus database 
INFO [08-01|21:09:11.518] Clearing DBs dirty flags 
INFO [08-01|21:09:11.522] Database set clean                       name=gossip
INFO [08-01|21:09:11.528] Database set clean                       name=lachesis
INFO [08-01|21:09:11.529] Archive state database reverted          block=968
INFO [08-01|21:09:11.529] Re-creating live state from the archive... 
INFO [08-01|21:09:11.768] Healing finished

sonicd --config=config.toml --datadir=/tmp/operatest-break --fakenet=0/2 --port 5051 --http --http.api=eth,debug,net,admin,web3,personal,txpool,ftm,dag --http.corsdomain * --verbosity=3 --nat extip:127.0.0.1
WARN [08-01|21:09:53.345] Please add '--cache 11968' flag to allocate more cache for Opera. Total memory is 23937 MB. 
INFO [08-01|21:09:53.346] Maximum peer count                       total=50
INFO [08-01|21:09:53.391] Genesis is written                       name=fake id=4003 genesis=0x4cbc03e802ab1d376d3983fc832a7ef8c7f1e5eb71ac558caaa0fe00427a3b7a
INFO [08-01|21:09:53.405] Loaded local transaction journal         transactions=2 dropped=2
INFO [08-01|21:09:53.405] Regenerated local transaction journal    transactions=0 accounts=0
INFO [08-01|21:09:53.415] Starting peer-to-peer node               instance=Sonic/v1.2.1-e/linux-amd64/go1.22.3
INFO [08-01|21:09:53.420] Mapped network port                      proto=tcp extport=5051 intport=5051 interface=ExtIP(127.0.0.1)
INFO [08-01|21:09:53.486] Mapped network port                      proto=udp extport=5051 intport=5051 interface=ExtIP(127.0.0.1)
INFO [08-01|21:09:53.488] New local node record                    seq=18 id=2982730356b1dbaa ip=127.0.0.1 udp=5051 tcp=5051
INFO [08-01|21:09:57.492] Started P2P networking                   self=enode://2cb2478b8637c9e26b1b3f9cd0590d47b95a6ec87d574ea8fcd64cda60e25fd6d4d8ae96788c87dfc48b7946f420dcaed7c7a9e1ab45d8dec266b0ed11eca18b@127.0.0.1:5051
INFO [08-01|21:09:57.493] IPC endpoint opened                      url=/tmp/operatest-break/opera.ipc
INFO [08-01|21:09:57.495] HTTP server started                      endpoint=127.0.0.1:18545 prefix= cors=* vhosts=localhost
INFO [08-01|21:09:58.513] New DAG summary                          new=1 last_id=19:1:a0b459 age=19m54.937s t="379.66µs"
INFO [08-01|21:09:58.514] New block                                index=969 id=19:3:352c42      gas_used=0 txs=0/0 age=19m53.012s t="244.268µs"
INFO [08-01|21:09:58.515] New block                                index=970 id=19:5:e9a49f      gas_used=0 txs=0/0 age=19m43.399s t="200.988µs"
INFO [08-01|21:09:58.515] New block                                index=971 id=19:7:f1acb0      gas_used=0 txs=0/0 age=19m33.786s t="149.792µs"
```

Known issues:
The healed database does not work good with emitter - emitting must be disabled until the node is synced, otherwise emitter may try to include block-records referencing events, which have been deleted - which lead to panic. We may consider deleting blocks too to avoid it, but emitter should be disabled until synced through. Probably affects only fakenet.

### CODE REVIEW

**Review by HerbertJordan — APPROVED** [review]
The Carmen integration looks good, about the remaining data bases I would like to ask you to get another reviewer's approval.

File: gossip/evmstore/carmen.go [hunk 1]
```diff
@@ -136,12 +137,20 @@ func (c *CarmenStateDB) HasSuicided(addr common.Address) bool {
 	return c.db.HasSuicided(cc.Address(addr))
 }
 
-func (c *CarmenStateDB) AddBalance(addr common.Address, amount *big.Int) {
-	c.db.AddBalance(cc.Address(addr), amount)
+func (c *CarmenStateDB) AddBalance(addr common.Address, amountInt *big.Int) {
+	am, err := amount.NewFromBigInt(amountInt)
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (thaarok) [reply gossip/evmstore/carmen.go:143]: ok, lets move the carmen upgrade into a standalone PR: https://github.com/Fantom-foundation/Sonic/pull/188

**Comment by thaarok** [comment]
Rebased on latest main, upgraded carmen the lastest, sucessfuly re-tested on local fakenet.

**Comment by pc-quiknode** [comment]
tried with production db without success

```
# sonictool --datadir=/sonicd --cache=15993 heal
INFO [08-06|23:02:39.570] Maximum peer count                       total=50
INFO [08-06|23:02:39.570] Forcing unlock of directory              dir=/sonicd/carmen/archive
failed to get checkpoint - probably none has been created in this database yet, healing not possible: open /sonicd/carmen/archive/roots/committed.json: no such file or directory
```

**Comment by thaarok** [comment]
> tried with production db without success
> 
> ```
> # sonictool --datadir=/sonicd --cache=15993 heal
> INFO [08-06|23:02:39.570] Maximum peer count                       total=50
> INFO [08-06|23:02:39.570] Forcing unlock of directory              dir=/sonicd/carmen/archive
> failed to get checkpoint - probably none has been created in this database yet, healing not possible: open /sonicd/carmen/archive/roots/committed.json: no such file or directory
> ```

The node needs to run some time with this patch to create a checkpoint. Then the heal allows to return back to the checkpoint.

This feature is in active development and needs to be optimized yet.

**Pull request merged**